### PR TITLE
Enhanced vCard QR Code Generator Script

### PR DESCRIPTION
**Pull request by mgiannopoulos24**
**Description:**

This pull request enhances the vCard QR Code Generator script with several improvements to functionality, usability, and maintainability:

- **Input Validation:** Added basic validation to ensure required fields are not empty before generating vCard data.
- **Error Handling:** Implemented error handling for exceptions during QR code generation to provide informative error messages.
- **Modular Structure:** Refactored the script into distinct functions (`get_vcard_data()` and `generate_qr_code(data)`) for clearer separation of concerns and improved code readability.
- **Documentation:** Updated the README with detailed installation instructions, usage guidance and an example interaction.

**Changes Made:**

- Added input validation to prevent empty required fields.
- Enhanced error handling for QR code generation.
- Refactored script into modular functions for improved maintainability.
- Updated README with installation instructions and usage guidance

These enhancements ensure the script handles various scenarios more robustly and provides clear instructions for users.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -1,14 +1,60 @@
 # vCard QR Code Generator
-
-Simple script to interactively generate a QR Code containing a vCard
+This is a simple Python script to interactively generate a QR Code containing vCard information. It utilizes the qrcode library to create the QR Code image.
```
Comment (ll-nick) [inline README.md]: This should be short and concise. I don't think naming dependencies should be done here. The first sentence is fine though.

File: README.md [hunk 2]
```diff
@@ -1,14 +1,60 @@
...
 
 ## Requirements
+- Python 3.x
+- qrcode Python library
```
Comment (ll-nick) [inline README.md]: Please link the library

File: README.md [hunk 3]
```diff
@@ -1,14 +1,60 @@
...
+2. **Install dependencies:**
+Make sure you have Python 3.x installed. Install the required `qrcode` library using pip:
+```console
+pip install qrcode
+```
 
 ## Usage
+Run the script from the command line:
+```
+python3 generate_vcard_qr_code.py
```
Comment (ll-nick) [inline README.md:23]: Fair enough, I guess this is more platform independent than using the shebang.

File: README.md [hunk 4]
```diff
@@ -1,14 +1,60 @@
...
+Organization: Example Inc.
+URL: https://example.com
+Email: john.doe@example.com
+Phone: +1234567890
+Street + number: 123 Main St
+City: Anytown
+Postal Code: 12345
+Country: USA
+vCard QR code generated and saved as 'vCardQR.png'.
 ```
```
Comment (ll-nick) [inline README.md]: Honestly, I think this example is a little over the top. The interaction should be simple enough. But having this verbose documentation means that more places have to be updated should the script ever change. So the added value vs the added cost of maintenance is not worth it.

File: generate_vcard_qr_code.py [hunk 5]
```diff
@@ -2,21 +2,27 @@
 
 import qrcode
 
-
```
Comment (ll-nick) [inline generate_vcard_qr_code.py:5]: I use the black formatter. There should be two empty lines after the imports.

File: generate_vcard_qr_code.py [hunk 6]
```diff
@@ -2,21 +2,27 @@
...
-    email = input("Email: ")
-    phone = input("Phone: ")
-    street = input("Street + number: ")
-    city = input("City: ")
-    postal_code = input("Postal Code: ")
-    country = input("Country: ")
-
+    
+    # Input collection with basic validation
+    first_name = input("First Name: ").strip()
```
Comment (ll-nick) [inline generate_vcard_qr_code.py:16]: Adding the strip() is a good idea 👍

File: generate_vcard_qr_code.py [hunk 7]
```diff
@@ -2,21 +2,27 @@
...
+    url = input("URL: ").strip()
+    email = input("Email: ").strip()
+    phone = input("Phone: ").strip()
+    street = input("Street + number: ").strip()
+    city = input("City: ").strip()
+    postal_code = input("Postal Code: ").strip()
+    country = input("Country: ").strip()
+
+    # Ensure required fields are not empty
+    if not (first_name and last_name and email and phone and street and city and postal_code and country):
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: If we are checking this, we should do it properly: 
Not all of these properties are strictly necessary. Check the vCard documentation for details. I'd suggest having errors or exceptions for these mandatory properties like you did. However, the optional ones should then just not be added to the vCard. Perhaps a warning log would be appropriate.

File: generate_vcard_qr_code.py [hunk 8]
```diff
@@ -30,12 +36,28 @@ def get_vcard_data():
...
-print("vCard QR code generated and saved as 'vCardQR.png'.")
+def generate_qr_code(data):
+    try:
+        # Create the QR code
+        qr = qrcode.QRCode(
+            version=1,
+            error_correction=qrcode.constants.ERROR_CORRECT_L,
+            box_size=10,
+            border=4,
+        )
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: If you deviate from the defaults of the library, you shouldn't hard code these in the function call but define these settings somewhere. Would also be nice to have these documented somewhere.

**Review by ll-nick — CHANGES_REQUESTED** [review]

**Comment by ll-nick** [comment]
Whoops, sent the review without a comment.

I didn't see a PR coming the way of this little script but I guess why not, so thanks for the contribution.

I commented on some of the changes, please take a look at my review and consider adjusting your PR or telling me why I'm wrong :D

File: README.md (hunk 1 above)
Comment (mgiannopoulos24) [reply README.md]: Fixed.

File: README.md (hunk 2 above)
Comment (mgiannopoulos24) [reply README.md]: Fixed.

File: README.md (hunk 4 above)
Comment (mgiannopoulos24) [reply README.md]: Fixed.

File: generate_vcard_qr_code.py (hunk 6 above)
Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py:16]: Thanks.

File: generate_vcard_qr_code.py (hunk 7 above)
Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: Fixed.

File: generate_vcard_qr_code.py (hunk 8 above)
Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: Absolutely, it's a good practice to define such settings in a configurable way, especially if they deviate from the library's defaults.

File: generate_vcard_qr_code.py [hunk 9]
```diff
@@ -1,41 +1,75 @@
...
+    first_name = input("First Name: ").strip()
+    last_name = input("Last Name: ").strip()
+    organization = input("Organization: ").strip()
+    url = input("URL: ").strip()
+    email = input("Email: ").strip()
+    phone = input("Phone: ").strip()
+    street = input("Street + number: ").strip()
+    city = input("City: ").strip()
+    postal_code = input("Postal Code: ").strip()
+    country = input("Country: ").strip()
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: The optional properties should be marked as optional in the user prompt. Alternatively (or both) the mandatory fields should be marked as such.

File: generate_vcard_qr_code.py [hunk 10]
```diff
@@ -1,41 +1,75 @@
...
-    organization = input("Organization: ")
-    url = input("URL: ")
-    email = input("Email: ")
-    phone = input("Phone: ")
-    street = input("Street + number: ")
-    city = input("City: ")
-    postal_code = input("Postal Code: ")
-    country = input("Country: ")
+    # Ensure required fields are not empty
+    if not (first_name and last_name and email and phone and street and city and postal_code and country):
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: This should only check for the mandatory properties.

File: generate_vcard_qr_code.py [hunk 11]
```diff
@@ -1,41 +1,75 @@
...
+        print("Error: Required fields cannot be empty.")
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: Optional idea: You could raise an exception here and replace the try catch block below by a single try catch block in the main for a more unified error handling.

File: generate_vcard_qr_code.py [hunk 12]
```diff
@@ -1,41 +1,75 @@
...
+    if email:
+        vcard_data += f"EMAIL;TYPE=work:{email}\n"
+    if phone:
+        vcard_data += f"TEL;TYPE=work:{phone}\n"
+    if street or city or postal_code or country:
+        vcard_data += f"ADR;TYPE=intl,work,postal,parcel:;;{street};{city};;{postal_code};{country}\n"
+
+    vcard_data += "END:VCARD"
 
     return vcard_data
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: Nice! :-)

**Review by ll-nick — CHANGES_REQUESTED** [review]
Thanks for the quick response! Looks way better, thanks for working in my feedback.

I found a couple of things to moan about, if you could take a look.

File: generate_vcard_qr_code.py (hunk 9 above)
Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: Made these optional.

File: generate_vcard_qr_code.py (hunk 11 above)
Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: Used this method to both check if the require fields are complete and raise error if not.

**Comment by mgiannopoulos24** [comment]
Check required:

```console
User@Github:~$ python3 generate_vcard_qr_code.py 
Please enter the following information for the vCard:
First Name: A
Last Name:
Organization: a
Error: First Name, Last Name, and Organization are required fields.
```

File: generate_vcard_qr_code.py [hunk 13]
```diff
@@ -1,41 +1,85 @@
...
+    try:
+        # Generate the vCard data
+        data = get_vcard_data()
 
-# Create the QR code
-img = qrcode.make(data)
-img.save("vCardQR.png")
+        if data:
+            # Create the QR code
+            generate_qr_code(data)
```
Comment (ll-nick) [inline generate_vcard_qr_code.py]: Alright, one last thing:
If you don't catch the exception above you can handle it here and don't need to return None and the check for that in the  "if data". 

So the generate_qr_code can raise an exception and the get_vcard_data can raise an exception and they are both handled in the same way by logging an error.

Returning None and then checking for that obfuscates what is actually going on, namely an exception.

**Review by ll-nick — CHANGES_REQUESTED** [review]
One last change and then I think we are good 👍

File: generate_vcard_qr_code.py (hunk 13 above)
Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: What do you think of this:

```python
if __name__ == "__main__":
    try:
        # Generate the vCard data
        data = get_vcard_data()
        
        # Create the QR code
        generate_qr_code(data)

    except ValidationError as ve:
        print(f"Validation error: {ve}")
    except Exception as e:
        print(f"Error: {e}")
```

Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: Or maybe you want it like so:

```python
def generate_qr_code(data):
    try:
        # Create the QR code with defined settings
        qr = qrcode.QRCode(
            version=QR_VERSION,
            error_correction=QR_ERROR_CORRECTION,
            box_size=QR_BOX_SIZE,
            border=QR_BORDER,
        )
        qr.add_data(data)
        qr.make(fit=True)

        img = qr.make_image(fill_color="black", back_color="white")
        img.save("vCardQR.png")
        print("vCard QR code generated and saved as 'vCardQR.png'.")
    except Exception as e:
        print(f"Error generating QR code: {e}")
        raise
```

Comment (ll-nick) [reply generate_vcard_qr_code.py]: Your first suggestion is exactly what I had in mind. Looks pretty clean, right?

Comment (mgiannopoulos24) [reply generate_vcard_qr_code.py]: Alright as you wish. Let me commit.

**Review by ll-nick — APPROVED** [review]
Thanks a lot for the contribution and the effort you put into adding my feedback. I'll merge this now 👍

**Comment by mgiannopoulos24** [comment]
Sure no worries. It was fun. :)

**Pull request merged**